client/alloc_runner.go

```diff
@@ -448,7 +454,11 @@ func (r *AllocRunner) Run() {
 			continue
 		}
 
-		tr := NewTaskRunner(r.logger, r.config, r.setTaskState, r.ctx, r.Alloc(), task.Copy(), r.vaultClient)
+		r.allocDirLock.Lock()
```

Comment:
Sadly we'd still need this lock because it guards the setting and checking the allocDir field on the AllocRunner.

We might be able to skip acquiring it on this particular line, but I'd rather always guard accesses to a field than try to figure out safe places to skip synchronization and confuse future readers of the code.